### ISSUE TITLE
feat: Add support for `.shorebirdignore` file

### ISSUE THREAD

**Issue by josiahsrc** [shorebirdtech/shorebird#2433]
**Description**

It's common for projects to have local files that aren't uploaded to git, e.g. `.env.local`. This creates a warning in shorebird saying that those files won't be included in the patch. 

```
[WARN] Your app contains asset changes, which will not be included in the patch.
    Added files:
        Products/Applications/Runner.app/Frameworks/App.framework/flutter_assets/assets/env/.env.emulators-local
    Changed files:
        Products/Applications/Runner.app/Frameworks/App.framework/flutter_assets/AssetManifest.bin
        Products/Applications/Runner.app/Frameworks/App.framework/flutter_assets/AssetManifest.json
        Products/Applications/Runner.app/Frameworks/App.framework/flutter_assets/fonts/MaterialIcons-Regular.otf
```

It would be helpful to be able to define a `.shorebirdignore` file and add specific files to exclude from warnings.

```.gitignore
# .shorebirdignore
assets/.env.emulators.local
```

**Requirements**

- [ ] Support adding `.shorebirdignore` file
- [ ] Exclude files specified in `.shorebirdignore` from warnings

**Additional Context**

In my scenario, I have a CI machine and a local dev machine that don't share the same local files. The CI machine creates the release, the local machine creates the patches. Because of this, the local asset files aren't shared.

Another related issue here might be that shorebird always thinks that the podfile.lock file has changed, but this may be because the CI machine and my machine have different xcode SHAs. Might be a separate issue, or something I'm doing wrong, but thought I'd mention it in case it's helpful.

```
[WARN] Your ios/Podfile.lock is different from the one used to build the release.
This may indicate that the patch contains native changes, which cannot be applied with a patch. Proceeding may result in unexpected behavior or crashes.
```

**Comment by bryanoltman** [comment]
This is an interesting idea. I think this is absolutely valid for assets (especially the `.env.emulators.local` example you gave), but I would want to make sure we don't hide real errors. For example, the podfile.lock case you mentioned is 100% *not* something we want people to ignore, as changes to plugin versions could cause your app to crash.

**Comment by bryanoltman** [comment]
It also looks like you're seeing warnings about the fonts changing, which is probably also a warning you should address: https://docs.shorebird.dev/troubleshooting/#i-see-a-your-app-contains-asset-changes-warning-when-running-shorebird-patch

**Comment by josiahsrc** [comment]
@bryanoltman  Agreed on the Podfile lock. It's strange because the first patch has no problem (the podfile lock is the same). But subsequent patches to the same release emit that warning for me, even when no changes have happened. I'll look more into this though, I see this is a common issue.

Thank you for linking me to the tree-shaking documentation, I'll be sure to add that in until asset patches are available.

**Comment by bryanoltman** [comment]
Do you have your pubspec.lock checked in? Especially on CI (and across different machines generally), it's very common for `flutter pub get` to pull down different package versions if run at different times if no pubspec.lock is present.

**Comment by eseidel** [comment]
Curious, how do you prevent `.env.emulators-local` from being included in you releases in your pre-shorebird flow?

**Comment by josiahsrc** [comment]
@bryanoltman Yes, my pubspec.lock is committed to git. I just noticed, however, that my Podfile.lock is not. I bet that's it; I'll add that in and see if that fixes the issue.

@eseidel I attempt to load in all `-local` env files on boot. Depending on if the files are present, the values are loaded into the program. The app runs normally whether they're included or not.

```dart
String fileContent;
try {
  fileContent = await rootBundle.loadString(fullPath, cache: false);
  fileContent = fileContent.trim();
} catch (e) {
  fileContent = '';
}
```

**Comment by eseidel** [comment]
I'd love to help, but I'm still not sure I understand the use case.

My sense is that you built the release (that you sent to the stores) from a machine that did not have a local-env file, correct?  (Or maybe built it from the same machine just intentionally removed the local-env file before releasing?)

And you'd like to build a patch from a machine that does include a local-env file, but not have that file included in your users apps after patching?  In my mind preparing a release vs a patch are very similar activities and so I don't immediately see why one would want to teach tools to ignore changes to local setup when building one but not the other.  But it's also 100% possible I'm missing something obvious (I didn't get my normal full dose of sleep last night). Thanks!

**Comment by josiahsrc** [comment]
@eseidel Apologies for the confusion (and get some rest!). Yes, that's correct. I'll try to explain my shorebird workflow a little more clearly below.

**Releases**
I build releases from a CI machine because it is a safe, unchanging process. It's very uncommon that things will go wrong with this pipeline and the release is slow enough that I have time to test on the different environments before it goes out (e.g. dev and prod).

**Patches**
Patches require a "surgical" procedure where I, as the developer, have to know specifically which version I need to change and I need to verify that the patch I'm producing is safe (checking for native dependency changes, asset changes, regressions, cherry-picked correctly, etc...). This exists outside of the normal release flow and is usually done in the case where a customer has reported a bug and I need to patch it on their specific version. Because of how involved this is from a QA/risk management perspective, it doesn't make sense to automate this.

Long story short, I haven't found a safe way to automate patches that doesn't end up taking more developer time to manage. This is why I run shorebird releases from the CI machine and shorebird patches from my local machine. Being able to exclude unwanted files from warnings, where I'm confident that it's not going to break things, would help remove a couple extra validation steps from this process.

**Comment by josiahsrc** [comment]
In an ideal world there would be one single source of truth version that everyone is on, but I don't believe that's possible in today's Flutter world

**Comment by eseidel** [comment]
Thank you, that's helpful.

If it helps, Shorebird does manage the Flutter version for you (to ensure it *exactly* matches what the release was made with).

In case it helps, aflow we've used it one of making "release branches" and then cherry-picking commits onto those release branches.  And then just setting up a builder to run `shorebird patch` on that branch (which should be otherwise identical to running `shorebird release`).  We should probably write up more about ways we've seen folks manage this.

**Comment by josiahsrc** [comment]
Good thought on the branching, that sounds like it would serve as a good substitute for patching locally. I'll give that a shot, thanks @eseidel 

There may still be some arguments for a `.shorebirdignore` file outside the scope of this thread, but please feel free to close this issue at your discretion.

**Comment by eseidel** [comment]
Thanks!  I think you raise an interesting issue here, but we'll just re-open when we get another request.  Thank you!  Please don't ever hesitate to reach out if we can help!
